onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc

```diff
@@ -86,9 +103,29 @@ class ONNX_OPERATOR_KERNEL_CLASS_NAME(kXnnpackExecutionProvider, kOnnxDomain, 13
 // Internal domain
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kXnnpackExecutionProvider, kDynamicDomainByCreate, 1, QLinearSoftmax);
 
-std::unique_ptr<KernelRegistry> RegisterKernels() {
-  auto kernel_registry = std::make_unique<onnxruntime::KernelRegistry>();
+#ifdef XNNPACK_FP16_SUPPORTED
+Status RegisterFp16Kernels(KernelRegistry& kernel_registry) {
+  static const BuildKernelCreateInfoFn function_table[] = {
+      BuildKernelCreateInfo<void>,  // default entry to avoid the list become empty after ops-reducing
+
+      KERNEL_CREATE_INFO_VERSIONED_TYPED(8, 9, MLFloat16, MaxPool, kMSInternalNHWCDomain),
+      KERNEL_CREATE_INFO_VERSIONED_TYPED(10, 10, MLFloat16, MaxPool, kMSInternalNHWCDomain),
+      KERNEL_CREATE_INFO_VERSIONED_TYPED(11, 11, MLFloat16, MaxPool, kMSInternalNHWCDomain),
+      KERNEL_CREATE_INFO_TYPED(12, MLFloat16, MaxPool, kMSInternalNHWCDomain),
+  };
```

Comment:
nit: what's the reason for not doing the simple thing and adding them inside an `#ifdef XNNPACK_FP16_SUPPORTED` at the end of the existing ones (after QLinearSoftmax)? A lot of the diffs here are to separate out the fp16 registration and I'm not sure what benefit that provides.

Reply:
Good question. I'll double check with Mlas guys why there're so many MLAS_F16VEC_INTRINSICS_SUPPORTED in CPU.

Reply:
I added some macros to make register F16 macros more easily